atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStoreTest.java

```diff
@@ -120,66 +138,217 @@ public void returnsTimestampAssociatedWithGreatestPunchedTimeEvenIfItIsNotGreate
     }
 
     @Test
-    public void getMillisForTimestampIfNotPunchedBeforeWhenPunchedExactlyAtLowerBound() {
+    public void findOlderReturnsReasonableEstimate() {
         KeyValueService kvs = Mockito.spy(new InMemoryKeyValueService(false));
-        puncherStore = initializePuncherStore(PUNCHER_HISTORY, kvs);
-        // Punched: (10, 100), (20, 200), (30, 300)
-        // Arguments: (10, 100)
-        // First ts punched at 100 is 10. 10 = 10, so we don't range scan and return 100.
-        assertThat(KeyValueServicePuncherStore.getMillisForTimestampIfNotPunchedBefore(kvs, TIMESTAMP_1, WALL_CLOCK_1))
-                .isEqualTo(WALL_CLOCK_1);
-        verify(kvs, times(1)).getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), eq(Long.MAX_VALUE));
-        verify(kvs, times(1)).getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), anyLong());
+        PuncherStore puncherStore = KeyValueServicePuncherStore.create(kvs);
+
+        int maxReads = 10;
+        long maxOffset = KeyValueServicePuncherStore.MAX_RANGE_SCAN_SIZE * LongMath.pow(2, maxReads - 1);
+        long minOffsetForNewerEntries = maxOffset * 3 / 4;
+
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        for (int i = 0; i < 100; i++) {
+            puncherStore.put(
+                    2_000 + random.nextLong(10_000L), random.nextLong(minOffsetForNewerEntries + 1, maxOffset));
+        }
+
+        for (int i = 0; i < 100; i++) {
+            puncherStore.put(random.nextLong(1_000L), random.nextLong(minOffsetForNewerEntries / 3));
+        }
+
+        MillisAndMaybeTimestamp result = KeyValueServicePuncherStore.findOlder(kvs, 1_000L, maxOffset);
+        assertThat(result.millis()).isLessThanOrEqualTo(maxOffset);
+        assertThat(result.timestamp()).hasValueSatisfying(ts -> assertThat(ts).isLessThanOrEqualTo(1_000L));
+        verify(kvs, atLeast(1)).getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), anyLong());
+        verify(kvs, atMost(maxReads)).getRange(eq(AtlasDbConstants.PUNCH_TABLE), any(RangeRequest.class), anyLong());
+    }
+
+    @Test
+    public void findOlderFuzzTest() {
+        KeyValueService kvs = Mockito.spy(new InMemoryKeyValueService(false));
+        PuncherStore puncherStore = KeyValueServicePuncherStore.create(kvs);
+
+        int maxReads = 10;
+        long maxOffset = KeyValueServicePuncherStore.MAX_RANGE_SCAN_SIZE * LongMath.pow(2, maxReads - 1);
+
+        ThreadLocalRandom random = ThreadLocalRandom.current();
+        for (int i = 0; i < 100; i++) {
+            puncherStore.put(random.nextLong(10_000L), random.nextLong(maxOffset));
+        }
+
+        MillisAndMaybeTimestamp result = KeyValueServicePuncherStore.findOlder(kvs, 1_000L, maxOffset);
+        assertThat(result.millis()).isLessThanOrEqualTo(maxOffset);
+        assertThat(result).satisfies(res -> res.timestampSatisfies(ts -> ts <= 1_000L));
```

Comment:
this could fail maybe around one in 10^10 times? not sure, if your random.nextLong(...) generates numbers such that you don't see anything smaller than 1000. In any case we should allow timestamp being absent anyway.